File: packages/core/src/xmn/xmn-core.ts
/*------------------------------------------------------------------------------
   About      : <Write about the file here>
   
   Created on : Thu Jun 29 2017
   Author     : Raghvendra Varma
   
   Copyright (c) 2017 Mubble Networks Private Limited. All rights reserved.
------------------------------------------------------------------------------*/

import { Mubble, RunContextBase } from '..'
import { CustomData }             from './custom-data'

export enum Protocol {HTTP, WEBSOCKET, HTTPS, HTTP_THIRD}

export const HANDSHAKE = '__handshake__'

/* HTTP Headers */
export namespace HTTP {

  // normally these keys are written with uppercase, we are writing them in lowercase 
  // for compatibility
  export const HeaderKey = {
    userAgent        : 'user-agent',
    clientSecret     : 'x-client-secret',
    contentType      : 'content-type',
    contentLength    : 'content-length',
    setCookie        : 'set-cookie',
    contentEncoding  : 'content-encoding',
    clientId         : 'x-obopay-cid',
    versionNumber    : 'x-obopay-version',
    requestTs        : 'x-obopay-ts',
    symmKey          : 'x-obopay-key',
    requestType      : 'x-obopay-type',
    bodyEncoding     : 'x-obopay-encoding',
    transferEncoding : 'transfer-encoding',
    location         : 'location',
    accept           : 'accept',
    authorization    : 'authorization',
    token            : 'token'
  }
  
  /* HTTP Headers */
  export const HeaderValue = {
    form     : 'application/x-www-form-urlencoded',
    mutiForm : 'multipart/form-data',
    stream   : 'application/octet-stream',
    json     : 'application/json',
    gzip     : 'gzip',
    deflate  : 'deflate',
    identity : 'identity',
    version2 : 'v2',
    chunked  : 'chunked'
  }

  export const Const = {
    protocolHttp  : 'http:',
    protocolHttps : 'https:'
  }

  export const Method = {
    PUT    : 'PUT',
    GET    : 'GET',
    POST   : 'POST',
    DELETE : 'DELETE'
  }

  export const CurrentProtocolVersion = HTTP.HeaderValue.version2
}

export const NetworkType = {
  net2G   : '2G',
  net3G   : '3G',
  net4G   : '4G',
  net5G   : '5G',
  wifi    : 'wifi',
  unknown : 'unk',
  absent  : 'absent'
}

export const WEB_SOCKET_URL = {
  ENC_PUBLIC    : 'socket.io',
  ENC_PRIVATE   : 'engine.io',
  PLAIN_PUBLIC  : 'rocket.io',
  PLAIN_PRIVATE : 'locket.io'
}

let lastReqId   : number = 0
let lastEventId : number = 0

export const WIRE_TYPE = {
  REQUEST     : 'REQUEST',
  EVENT       : 'EVENT',
  SYS_EVENT   : 'SYS_EVENT',
  EPH_EVENT   : 'EPH_EVENT',
  EVENT_RESP  : 'EVENT_RESP',
  REQ_RESP    : 'REQ_RESP'
}

export class WireObject {

  static getWireObject(json: any) {
    switch (json.type) {
      
      case WIRE_TYPE.REQUEST:
      return new WireRequest(json.name, json.data, json.ts)

      case WIRE_TYPE.EVENT:
      return new WireEvent(json.name, json.data, json.ts)

      case WIRE_TYPE.EPH_EVENT:
      return new WireEphEvent(json.name, json.data, json.ts)

      case WIRE_TYPE.SYS_EVENT:
      return new WireSysEvent(json.name, json.data)

      case WIRE_TYPE.EVENT_RESP:
      return new WireEventResp(json.name, json.ts, json.data, json.errorCode, json.errorMessage)

      case WIRE_TYPE.REQ_RESP:
      return new WireReqResp(json.name, json.ts, json.data, json.errorCode, json.errorMessage)

      default:
      console.info('Error: Invalid wire object ' + JSON.stringify(json))
      return null
    }
  }

  static parseString(str : string) : WireObject {
    return JSON.parse(str)
  }

  type      : string
  name      : string
  ts        : number
  data      : Mubble.uObject<any>

  constructor(type: string, name: string, data: object, ts ?: any) {
    this.type = type
    this.name = name
    this.data = data
    this.ts   = ts || Date.now() * 1000
  }

  stringify(): string {
    return JSON.stringify(this, (key, value) => 
      key.startsWith('_') || 
      (value && value.constructor.hasOwnProperty('byteLength')) ? undefined : value)
  }

  toString() {

    const err = (this as any).error

    return this.type + ':' + this.name + '@' + this.ts + ' ' + 
           (err || JSON.stringify(this.data).substr(0, 50))
  }
}

export class WireRequest extends WireObject {

  _isSent    : boolean = false
  constructor(apiName: string, data: object, ts ?: number, 
              public resolve ?:any, public reject ?:any) {
    super(WIRE_TYPE.REQUEST, apiName, data, ts)
    if (!ts) {
      if (this.ts === lastReqId) this.ts++
      lastReqId = this.ts
    }
  }
}

export class WireEvent extends WireObject {

  constructor(eventName: string, data: object, ts ?: number) {

    super(WIRE_TYPE.EVENT, eventName, data, ts)

    if (!ts) {
      if (this.ts === lastEventId) this.ts++
      lastEventId = this.ts
    }
  }
}

export class WireEphEvent extends WireObject {

  constructor(eventName: string, data: object, ts ?: number) {
    super(WIRE_TYPE.EPH_EVENT, eventName, data, ts)
  }
}

export class WireReqResp extends WireObject {
  errorCode    : string | null
  errorMessage : string | null
  errorObject  : Mubble.uObject<any> | undefined

  _err ?: any   // Full Error Object Instance. need not go to client (_). Required for trace logging
  constructor(name : string, ts : number, data : object, errorCode ?: string, errorMessage ?: string,
              errorObject ?: Mubble.uObject<any>, fullErr ?: any) {

    super(WIRE_TYPE.REQ_RESP, name, data, ts)

    this.errorCode    = errorCode || null
    this.errorMessage = errorMessage || null
    this.errorObject  = errorObject
    this._err         = fullErr
  }
}

export class WireEventResp extends WireObject {
  errorCode    : string | null
  errorMessage : string | null
  errorObject  : Mubble.uObject<any> | undefined

  _err ?: any   // Full Error Object Instance. need not go to client (_). Required for trace logging
  constructor(name : string, ts : number, data ?: object, errorCode ?: string, errorMessage ?: string,
              errorObject ?: Mubble.uObject<any>, fullErr ?: any) {
    
    super(WIRE_TYPE.EVENT_RESP, name, data || {}, ts)

    this.errorCode    = errorCode || null
    this.errorMessage = errorMessage || null
    this.errorObject  = errorObject
    this._err         = fullErr
  }
}

export const SYS_EVENT = {
  WS_PROVIDER_CONFIG      : 'WS_PROVIDER_CONFIG',
  ERROR                   : 'ERROR',
  PING                    : 'PING'
}

export class WireSysEvent extends WireObject {

  constructor(eventName: string, data: object) {
    super(WIRE_TYPE.SYS_EVENT, eventName, data)
  }
}

export interface WebSocketConfig {
  msPingInterval   : number
  syncKey         ?: string
}

export interface WssProviderConfig {
  pingSecs      : number
  maxOpenSecs   : number
  toleranceSecs : number
  key           : string
  custom        : CustomData
}

export const WssErrorCode = {
  HANDSHAKE_FAILURE : 501,
  INVALID_REQUESTTS : 502
}

export interface ConnectionError {
  code : string
  msg  : string
}

export interface InvocationData {
  name    : string
  ts      : number
  params  : object
}

export const DataLeader = {
  BINARY       : 0x01,
  DEF_JSON     : 0x02,
  JSON         : 0x03,
  ENC_BINARY   : 0x04,
  ENC_DEF_JSON : 0x05,
  ENC_JSON     : 0x06
}

export const Encoder = {
  MIN_SIZE_TO_COMPRESS  : 500
}

export interface XmnProvider {
  send(rc: RunContextBase , data: WireObject[]) : void
  requestClose(rc : RunContextBase) : void
}

export interface ActiveProviderCollection {
  addActiveProvider(clientId : number, provider : XmnProvider) : void
  getActiveProvider(clientId : number) : XmnProvider | undefined
}

File: packages/android/java/xmn/xmn-core.kt
package xmn

import core.JsonSerializable
import org.json.JSONArray
import org.json.JSONObject

/**
 * Created by
 * siddharthgarg on 04/04/18.
 */

var lastReqId   : Long = 0L
var lastEventId : Long = 0L

const val PROTOCOL_VERSION    = "v2"
const val PROTOCOL_HANDSHAKE  = "__handshake__"

object WireType {
  const val REQUEST     = "REQUEST"
  const val EVENT       = "EVENT"
  const val SYS_EVENT   = "SYS_EVENT"
  const val EPH_EVENT   = "EPH_EVENT"
  const val EVENT_RESP  = "EVENT_RESP"
  const val REQ_RESP    = "REQ_RESP"
}

object SysEvent {
  const val WS_PROVIDER_CONFIG      = "WS_PROVIDER_CONFIG"
  const val ERROR                   = "ERROR"
  const val PING                    = "PING"
}

object Encoder  {
  const val MIN_SIZE_TO_COMPRESS = 500
}

@Suppress("unused")
object DataLeader {

  const val BINARY       = 0x01
  const val DEF_JSON     = 0x02
  const val JSON         = 0x03
  const val ENC_BINARY   = 0x04
  const val ENC_DEF_JSON = 0x05
  const val ENC_JSON     = 0x06
}

@Suppress("unused")
object XmnError {

  const val errorCode           = 555

  const val NetworkNotPresent   = "NetworkNotPresent"  // Network is absent
  const val ConnectionFailed    = "ConnectionFailed"   // server connect problem= server not running, no network, connection break
  const val RequestTimedOut     = "RequestTimedOut"    // ideally means api bug
  const val SendTimedOut        = "SendTimedOut"       // ideally means terribly slow connection
  const val UnAuthorized        = "UnAuthorized"       // When the client id is not valid (server to server)

  const val _ConnectionExpired  = "_ConnectionExpired"
  const val _NotReady           = "_NotReady"
}

open class WireObject(val type: String, val name: String, var data: Any,
                      open var ts: Long = System.currentTimeMillis() * 1000): JsonSerializable {

  companion object {

    fun getWireObject(json: JSONObject): Any? {

      val type          = json.getString("type")
      val name          = json.getString("name")
      val data          = json.get("data")
      val ts            = json.getLong("ts")
      val errorCode     = json.optString("errorCode", "")
      val errorMessage  = json.optString("errorMessage", "")

      return when (type) {

        WireType.REQUEST    -> WireRequest(name, data, ts)
        WireType.SYS_EVENT  -> WireSysEvent(name, data)
        WireType.REQ_RESP   -> WireReqResp(name, data, ts, errorCode, errorMessage)
        WireType.EPH_EVENT  -> WireEphEvent(name, data, ts)
        WireType.EVENT      -> WireEvent(name, data, ts)
        WireType.EVENT_RESP -> WireEventResp(name, data, ts, errorCode, errorMessage)
        else                 -> null
      }
    }
  }

  override fun toJsonObject(): JSONObject {

    val json = JSONObject()
    json.put("type", type)
    json.put("name", name)
    json.put("data", data)
    json.put("ts", ts)

    return json
  }

  fun stringify(): String {
    return toJsonObject().toString()
  }
}

class WireRequest(apiName: String, data: Any, override var ts: Long):
    WireObject(WireType.REQUEST, apiName, data, ts) {

  var isSent: Boolean = false

  init {
    if (ts <= 0) {
      if (ts == lastReqId) ts = lastReqId + 1
      lastReqId = ts
    }
  }
}

class WireReqResp(name: String, data: Any, ts: Long, val errorCode: String? = null, val errorMessage: String? = null) :
    WireObject(WireType.REQ_RESP, name, data, ts), JsonSerializable {

  override fun toJsonObject(): JSONObject {

    val json = super.toJsonObject()
    if (!errorCode.isNullOrBlank()) json.put("errorCode", errorCode)
    if (!errorMessage.isNullOrBlank()) json.put("errorMessage", errorMessage)
    return json
  }
}

class WireEventResp(name: String, data: Any?, ts: Long, val errorCode: String? = null,val errorMessage: String? = null):
    WireObject(WireType.EVENT_RESP, name, data?:JSONObject(), ts), JsonSerializable {

  override fun toJsonObject(): JSONObject {

    val json = super.toJsonObject()
    if (!errorCode.isNullOrBlank()) json.put("errorCode", errorCode)
    if (!errorMessage.isNullOrBlank()) json.put("errorMessage", errorMessage)
    return json
  }
}

class WireEvent(eventName: String, data: Any, override var ts: Long):
    WireObject(WireType.EVENT, eventName, data, ts) {

  init {
    if (ts <= 0) {
      if (ts == lastEventId) ts = lastEventId + 1
      lastEventId = ts
    }
  }
}

class WireSysEvent(name: String, data: Any)
  : WireObject(WireType.SYS_EVENT, name, data)

class WireEphEvent(eventName: String, data: Any, ts : Long)
  : WireObject(WireType.EPH_EVENT, eventName, data, ts)

data class WsProviderConfig(var pingSecs: Int, var maxOpenSecs: Int, var toleranceSecs: Int,
                            var key: String?, var custom: CustomData?): JsonSerializable {

  constructor(json: JSONObject) : this(json.optInt("pingSecs"), json.optInt("maxOpenSecs"),
                                       json.optInt("toleranceSecs"), json.optString("key"),
                                       if (json.optJSONObject("custom") != null) CustomData.fromJsonObject(json.optJSONObject("custom")) else null)

  override fun toJsonObject(): JSONObject {

    val obj = JSONObject()
    obj.put("pingSecs", pingSecs)
    obj.put("maxOpenSecs", maxOpenSecs)
    obj.put("toleranceSecs", toleranceSecs)
    obj.put("key", key)
    obj.put("custom", custom?.toJsonObject())

    return obj
  }
}

@Suppress("unused")
class ConnectionError(data: Any) {
  val json  : JSONObject  = data as JSONObject
  val code  : String      = json.optString("code", "")
  val msg   : String      = json.optString("msg", "")
}

class RouterResponse(val errorCode: String?, val errorMessage: String?, val data: Any? = null): JsonSerializable {

  private var events : MutableList<WireObject> = mutableListOf()

  fun addEvents(events: MutableList<WireObject>) {
    this.events.addAll(events)
  }

  override fun toJsonObject(): JSONObject {

    val eventsArr = JSONArray()
    events.forEach {
      eventsArr.put(it.toJsonObject())
    }

    val json = JSONObject()
    json.put("errorCode", errorCode)
    json.put("errorMessage", errorMessage)
    json.put("data", data)
    json.put("events", eventsArr)

    return json
  }
}

interface XmnProvider {
  fun send(data: Array<WireObject>): String?
}
